internal/pkg/tokenbucket/token_bucket.go

```diff
@@ -0,0 +1,93 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package tokenbucket
```

Comment:
Could we use a preexisting implementation? https://pkg.go.dev/golang.org/x/time/rate perhaps?

https://pkg.go.dev/golang.org/x/time/rate#Limiter specifically is a token bucket rate limiter.

Disclaimer that I've never used this so it might be unsuitable.

Reply:
i don't have an experience with that package as well but it seems a bit like overkill

Reply:
The API is more complex, but the advantage is we don't have to maintain it and given it is a golang.org/x package we can trust it to generally be correct and kept up to date.

Reply:
I think it is worth trying to replace your implementation with this one, I suspect the maintenance cost of it will be lower over time if there's any chance we start adding features to this tokenbucket package in the future.